### PR TITLE
#78 Add support for customizable `bind_to_entity`

### LINKED ISSUE #78: Allow Entity objects to customize their contribution to Entity definition
Each object in `Entity` class definition, like `Meta`, `Field`s, etc., should be allowed to define how they will contribute to the Entity's structure.

This is a precursor requirement to implement related objects because a relation will be identified by an attribute name but have a shadow attribute tracking the target object ID.

### PR DESCRIPTION
**Pull request by subhashb**
All Entity attributes should implement a `bind_to_entity` to define how they will contribute to the Entity class. Some attributes, like Linked fields (Related objects), could have custom implementations that dynamically add a shadow attribute to the Entity, to represent a primary key of the target object.

Closes #78

### CODE REVIEW

**Comment by subhashb** [comment]
This PR also includes changes to rename `meta_` to `_meta`

File: src/protean/core/entity.py [hunk 1]
```diff
@@ -11,71 +11,100 @@
...
+ENTITY_META_ATTRIBUTE_NAMES = ('ordering')
+
 
 class EntityBase(type):
     """
     This base metaclass sets a `_meta` attribute on the Entity to an instance
     of Meta defined in the Entity
 
-    It also sets dictionary named `declared_fields` on the `meta_` attribute.
+    It also sets dictionary named `declared_fields` on the `_meta` attribute.
```
Comment (abhishek-ram) [inline src/protean/core/entity.py]: we cannot do this, as this makes `_meta` a private member and thus inaccessible for use outside

Comment (subhashb) [reply src/protean/core/entity.py]: Yeah, need to do one more thing here of exposing meta as `opts` or something similar, like Django does. I can do that next in a different PR.

This was optional, but I realized a full-blown `EntityMeta` implementation is necessary to be able to implement linked fields. Which is necessary to support hierarchy in entities and attribute overriding.

Comment (abhishek-ram) [reply src/protean/core/entity.py]: Hmm okay please expose it then

File: src/protean/core/entity.py [hunk 2]
```diff
@@ -11,71 +11,100 @@
...
+        new_class.add_to_class('_meta', EntityMeta(meta))
+
+        # Add declared fields
+        declared_fields = []
         for attr_name, attr_obj in attrs.items():
             if isinstance(attr_obj, Field):
-                # Bind the field object and append to list
-                attr_obj.bind_to_entity(klass, attr_name)
-                fields.append((attr_name, attr_obj))
+                new_class.add_to_class(attr_name, attr_obj)
```
Comment (abhishek-ram) [inline src/protean/core/entity.py]: this needs to be outside if as we want to add to class for all attributes

Comment (subhashb) [reply src/protean/core/entity.py]: Good catch. Will change and push as part of next commit.

Comment (subhashb) [reply src/protean/core/entity.py]: Pushed in later commit

**Pull request merged**